// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: email/v1/email_verification_service.proto

package emailpb

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on SendVerificationCodeRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SendVerificationCodeRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SendVerificationCodeRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SendVerificationCodeRequestMultiError, or nil if none found.
func (m *SendVerificationCodeRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SendVerificationCodeRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetEmailAddress() == nil {
		err := SendVerificationCodeRequestValidationError{
			field:  "EmailAddress",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetEmailAddress()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SendVerificationCodeRequestValidationError{
					field:  "EmailAddress",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SendVerificationCodeRequestValidationError{
					field:  "EmailAddress",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetEmailAddress()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SendVerificationCodeRequestValidationError{
				field:  "EmailAddress",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.GetAuth() == nil {
		err := SendVerificationCodeRequestValidationError{
			field:  "Auth",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetAuth()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SendVerificationCodeRequestValidationError{
					field:  "Auth",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SendVerificationCodeRequestValidationError{
					field:  "Auth",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetAuth()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SendVerificationCodeRequestValidationError{
				field:  "Auth",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return SendVerificationCodeRequestMultiError(errors)
	}

	return nil
}

// SendVerificationCodeRequestMultiError is an error wrapping multiple
// validation errors returned by SendVerificationCodeRequest.ValidateAll() if
// the designated constraints aren't met.
type SendVerificationCodeRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SendVerificationCodeRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SendVerificationCodeRequestMultiError) AllErrors() []error { return m }

// SendVerificationCodeRequestValidationError is the validation error returned
// by SendVerificationCodeRequest.Validate if the designated constraints
// aren't met.
type SendVerificationCodeRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SendVerificationCodeRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SendVerificationCodeRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SendVerificationCodeRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SendVerificationCodeRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SendVerificationCodeRequestValidationError) ErrorName() string {
	return "SendVerificationCodeRequestValidationError"
}

// Error satisfies the builtin error interface
func (e SendVerificationCodeRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSendVerificationCodeRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SendVerificationCodeRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SendVerificationCodeRequestValidationError{}

// Validate checks the field values on SendVerificationCodeResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SendVerificationCodeResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SendVerificationCodeResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SendVerificationCodeResponseMultiError, or nil if none found.
func (m *SendVerificationCodeResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *SendVerificationCodeResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Result

	if len(errors) > 0 {
		return SendVerificationCodeResponseMultiError(errors)
	}

	return nil
}

// SendVerificationCodeResponseMultiError is an error wrapping multiple
// validation errors returned by SendVerificationCodeResponse.ValidateAll() if
// the designated constraints aren't met.
type SendVerificationCodeResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SendVerificationCodeResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SendVerificationCodeResponseMultiError) AllErrors() []error { return m }

// SendVerificationCodeResponseValidationError is the validation error returned
// by SendVerificationCodeResponse.Validate if the designated constraints
// aren't met.
type SendVerificationCodeResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SendVerificationCodeResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SendVerificationCodeResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SendVerificationCodeResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SendVerificationCodeResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SendVerificationCodeResponseValidationError) ErrorName() string {
	return "SendVerificationCodeResponseValidationError"
}

// Error satisfies the builtin error interface
func (e SendVerificationCodeResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSendVerificationCodeResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SendVerificationCodeResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SendVerificationCodeResponseValidationError{}

// Validate checks the field values on CheckVerificationCodeRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CheckVerificationCodeRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CheckVerificationCodeRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CheckVerificationCodeRequestMultiError, or nil if none found.
func (m *CheckVerificationCodeRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CheckVerificationCodeRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetEmailAddress() == nil {
		err := CheckVerificationCodeRequestValidationError{
			field:  "EmailAddress",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetEmailAddress()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CheckVerificationCodeRequestValidationError{
					field:  "EmailAddress",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CheckVerificationCodeRequestValidationError{
					field:  "EmailAddress",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetEmailAddress()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CheckVerificationCodeRequestValidationError{
				field:  "EmailAddress",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.GetCode() == nil {
		err := CheckVerificationCodeRequestValidationError{
			field:  "Code",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetCode()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CheckVerificationCodeRequestValidationError{
					field:  "Code",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CheckVerificationCodeRequestValidationError{
					field:  "Code",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCode()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CheckVerificationCodeRequestValidationError{
				field:  "Code",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.GetAuth() == nil {
		err := CheckVerificationCodeRequestValidationError{
			field:  "Auth",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetAuth()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CheckVerificationCodeRequestValidationError{
					field:  "Auth",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CheckVerificationCodeRequestValidationError{
					field:  "Auth",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetAuth()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CheckVerificationCodeRequestValidationError{
				field:  "Auth",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CheckVerificationCodeRequestMultiError(errors)
	}

	return nil
}

// CheckVerificationCodeRequestMultiError is an error wrapping multiple
// validation errors returned by CheckVerificationCodeRequest.ValidateAll() if
// the designated constraints aren't met.
type CheckVerificationCodeRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CheckVerificationCodeRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CheckVerificationCodeRequestMultiError) AllErrors() []error { return m }

// CheckVerificationCodeRequestValidationError is the validation error returned
// by CheckVerificationCodeRequest.Validate if the designated constraints
// aren't met.
type CheckVerificationCodeRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CheckVerificationCodeRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CheckVerificationCodeRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CheckVerificationCodeRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CheckVerificationCodeRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CheckVerificationCodeRequestValidationError) ErrorName() string {
	return "CheckVerificationCodeRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CheckVerificationCodeRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCheckVerificationCodeRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CheckVerificationCodeRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CheckVerificationCodeRequestValidationError{}

// Validate checks the field values on CheckVerificationCodeResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CheckVerificationCodeResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CheckVerificationCodeResponse with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// CheckVerificationCodeResponseMultiError, or nil if none found.
func (m *CheckVerificationCodeResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CheckVerificationCodeResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Result

	if len(errors) > 0 {
		return CheckVerificationCodeResponseMultiError(errors)
	}

	return nil
}

// CheckVerificationCodeResponseMultiError is an error wrapping multiple
// validation errors returned by CheckVerificationCodeResponse.ValidateAll()
// if the designated constraints aren't met.
type CheckVerificationCodeResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CheckVerificationCodeResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CheckVerificationCodeResponseMultiError) AllErrors() []error { return m }

// CheckVerificationCodeResponseValidationError is the validation error
// returned by CheckVerificationCodeResponse.Validate if the designated
// constraints aren't met.
type CheckVerificationCodeResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CheckVerificationCodeResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CheckVerificationCodeResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CheckVerificationCodeResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CheckVerificationCodeResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CheckVerificationCodeResponseValidationError) ErrorName() string {
	return "CheckVerificationCodeResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CheckVerificationCodeResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCheckVerificationCodeResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CheckVerificationCodeResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CheckVerificationCodeResponseValidationError{}
